matchbox_server/src/topology.rs

```diff
@@ -1,277 +1,155 @@
-        let peer = self.clients.remove(peer_id);
-
-        if let Some(ref peer) = peer {
-            // Best effort to remove peer from their room
-            _ = self
-                .rooms
-                .get_mut(&peer.room)
-                .map(|room| room.remove(peer_id));
-        }
-        peer
-    }
-
-    /// Send a message to a peer without blocking.
-    fn try_send(&self, id: PeerId, message: Message) -> Result<(), ServerError> {
-        let peer = self.clients.get(&id);
-        let peer = match peer {
-            Some(peer) => peer,
-            None => {
-                return Err(ServerError::UnknownPeer);
-            }
-        };
-
-        peer.sender.send(Ok(message)).map_err(ServerError::from)
-    }
-}
-
-/// The handler for the HTTP request to upgrade to WebSockets.
-/// This is the last point where we can extract TCP/IP metadata such as IP address of the client.
-pub(crate) async fn ws_handler(
-    ws: WebSocketUpgrade,
-    room_id: Option<Path<RoomId>>,
-    Query(params): Query<HashMap<String, String>>,
-    State(state): State<Arc<Mutex<ServerState>>>,
-    ConnectInfo(addr): ConnectInfo<SocketAddr>,
-) -> impl IntoResponse {
-    info!("`{addr}` connected.");
-
-    let room_id = room_id.map(|path| path.0).unwrap_or_default();
-    let next = params
-        .get("next")
-        .and_then(|next| next.parse::<usize>().ok());
-
-    // Finalize the upgrade process by returning upgrade callback to client
-    ws.on_upgrade(move |websocket| handle_ws(websocket, state, RequestedRoom { id: room_id, next }))
-}
-
-fn parse_request(request: Result<Message, Error>) -> Result<JsonPeerRequest, ClientRequestError> {
-    let request = request?;
-
-    let request = match request {
-        Message::Text(text) => JsonPeerRequest::from_str(&text)?,
-        Message::Close(_) => return Err(ClientRequestError::Close),
-        _ => return Err(ClientRequestError::UnsupportedType),
-    };
-
-    Ok(request)
-}
-
-fn spawn_sender_task(
-    sender: SplitSink<WebSocket, Message>,
-) -> mpsc::UnboundedSender<Result<Message, Error>> {
-    let (client_sender, receiver) = mpsc::unbounded_channel();
-    tokio::task::spawn(UnboundedReceiverStream::new(receiver).forward(sender));
-    client_sender
-}
-
-/// One of these handlers is spawned for every web socket.
-async fn handle_ws(
-    websocket: WebSocket,
-    state: Arc<Mutex<ServerState>>,
-    requested_room: RequestedRoom,
-) {
-    let (ws_sender, mut ws_receiver) = websocket.split();
-    let sender = spawn_sender_task(ws_sender);
-
-    let peer_uuid = uuid::Uuid::new_v4().into();
-
-    {
-        let mut add_peer_state = state.lock().await;
-
-        let peers = add_peer_state.add_peer(Peer {
-            uuid: peer_uuid,
+#[derive(Debug, Default)]
+pub struct MatchmakingDemoTopology;
+
+#[async_trait]
+impl SignalingTopology<NoCallbacks, ServerState> for MatchmakingDemoTopology {
+    async fn state_machine(upgrade: WsStateMeta<NoCallbacks, ServerState>) {
```

Comment:
Feels like a lot of this could be mopved to `matchbox_signaling` if we were to implement a peer-to-peer `Room` type

Reply:
I don't like the idea of including a custom concept like "Room" in the matchbox signaling API, personally. It feels a little opinionated, but I've always had this reservation.

This is exactly why a custom topology is necessary in this case, but it kindof helps a bit since we don't have any examples of custom topologies. The demo helps with that, since it is one.

Reply:
Ah, it's been too long since I looked at that code, what I meant by room is exactly what implementations of `SignalingTopology` do :facepalm:. And in my ideal world we have topology wrappers like the aforementioned `ByName` and `ByCapacity` giving us the ability to construct what's currently implemented as `MatchmakingDemoTopology` as `ByName<ByCapacity<FullMesh>>`

Reply:
Oh, I see. I both like and dislike this because these wrappers can quickly become complex.

`ByName<ByCapacity<ByUuid<BySocketAddr<ByBearerAuthentication<...>>>>`

Maybe better is that we do what Bevy does with `SystemParam`. Essentially make your own function to handle on_connection and pass it, with any number of parameters/extractors that impl SystemParam.

Anywho, this is a bit out of scope for this PR. But I like the idea!